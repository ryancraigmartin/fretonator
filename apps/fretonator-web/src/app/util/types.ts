import { ScaleDegrees } from './constants';

export interface Fret {
  string: string;
  fret: number;
  displayName: string;
  degree: ScaleDegrees | string;
  degreeNumber?: number;
}

export enum ChordType {
  major = 'maj',
  minor = 'min',
  diminished = 'dim',
  augmented = 'aug'
}

export interface Chord {
  note: string;
  type: ChordType;
  romanNumeral: string;
}

export type FretMap = Fret[];

export type ChordMap = Chord[];

export enum NoteSymbol {
  aFlat = 'a_',
  a = 'a',
  bFlat = 'b_',
  b = 'b',
  c = 'c',
  cSharp = 'c#',
  dFlat = 'd_',
  d = 'd',
  dSharp = 'd#',
  eFlat = 'e_',
  e = 'e',
  f = 'f',
  fSharp = 'f#',
  gFlat = 'g_',
  g = 'g',
  gSharp = 'g#',
}

export enum Mode {
  ionian = 'ionian',
  dorian = 'dorian',
  phrygian = 'phrygian',
  lydian = 'lydian',
  mixolydian = 'mixolydian',
  aeolian = 'aeolian',
  locrian = 'locrian',
  harmonicMinor = 'harmonicMinor',
  phrygianDominant = 'phrygianDominant',
  aeolianDominant = 'aeolianDominant',
  majorPentatonic = 'majorPentatonic',
  minorPentatonic = 'minorPentatonic'
}

export interface JamTrack {
  key: NoteSymbol;
  mode: Mode;
  youTubeId: string;
  title: string;
  creator: string;
}

export type JamTracks = JamTrack[];

export interface NoteObject {
  name: string;
  flat: boolean;
  sharp: boolean;
  doubleFlat: boolean;
  doubleSharp: boolean;
  displayName?: string;
  degree?: string;
}

export type ModeMap = NoteObject[];

export enum NoteExtenderString {
  sharp = 'sharp',
  natural = 'natural',
  flat = 'flat'
}

export enum NoteExtenderSymbol {
  sharp = '#',
  natural = '',
  flat = '_'
}

export interface SimilarMode {
  noteDisplayName: string,
  note: string,
  noteExtender: NoteExtenderString;
  mode: Mode,
}

export type SimilarModes = SimilarMode[];
