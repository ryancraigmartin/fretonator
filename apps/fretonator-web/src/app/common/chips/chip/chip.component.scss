@import '../../../../styles/mixins';

:host {
  margin: 0.25rem;
  display: flex;
}

.chip {
  @include chip_button_base();
}

.chip--muted {
  border-color: var(--chip-border-color-muted);
  color: var(--chip-foreground-color-muted);
  background-color: var(--chip-background-color-muted);
}

.chip--round {
  height: 2.75rem;
  width: 2.75rem;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.chip--active {
  background-color: var(--chip-background-color-active);
  color: var(--chip-foreground-color-active);
  border-color: var(--chip-border-color-active);

  &.chip--muted {
    background-color: var(--chip-background-color-muted-active);
    border-color: var(--chip-border-color-muted-active);
    color: var(--chip-foreground-color-muted-active);
  }
}
