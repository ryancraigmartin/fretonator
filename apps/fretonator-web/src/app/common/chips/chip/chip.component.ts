import { Component, DoCheck, EventEmitter, Input, Output } from '@angular/core';
import { Mode } from '../../../util/types';
import { ActivatedRoute } from '@angular/router';

export enum SelectedColor {
  default = 'default',
  muted = 'muted',
}

@Component({
  selector: 'app-chip',
  templateUrl: './chip.component.html',
  styleUrls: ['./chip.component.scss']
})
export class ChipComponent implements DoCheck {
  @Input() note!: string;
  @Input() noteExtender!: string;
  @Input() mode!: Mode;
  @Input() selectedColor: 'default' | 'muted' = SelectedColor.default;
  @Input() rounded = false;
  @Input() isBaseNote = false;
  @Output() chipClick = new EventEmitter<MouseEvent>();
  SelectedColor = SelectedColor;
  selected = false;

  constructor(private activatedRoute: ActivatedRoute) {
  }

  ngDoCheck() {
    this.selected = this.isSelected(this.activatedRoute);
  }

  isSelected(activatedRoute: ActivatedRoute) {
    const data = activatedRoute.snapshot.data['selected'];

    if (data.note !== this.note) {
      return false;
    } else if (data.note === this.note && this.isBaseNote) {
      return true;
    }

    if (data.noteExtender !== this.noteExtender) {
      return false;
    }

    if (data.mode !== this.mode) {
      return false;
    }

    return true;
  }

  onClick(event: MouseEvent) {
    this.chipClick.emit(event);
  }
}
