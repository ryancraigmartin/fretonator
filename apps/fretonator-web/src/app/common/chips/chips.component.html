<div class="chips chips--rounded">
  <ng-content select="app-chip"></ng-content>
</div>
