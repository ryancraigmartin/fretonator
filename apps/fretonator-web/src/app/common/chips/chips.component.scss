@import '../../../styles/mixins';

.chips {
  @include chips_group;
}

.chips--rounded {
  margin-top: 0.75rem;
}
