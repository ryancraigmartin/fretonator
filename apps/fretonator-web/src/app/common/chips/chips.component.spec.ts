import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { ChipsModule } from './chips.module';
import { SelectedColor } from './chip/chip.component';
import { RouterTestingModule } from '@angular/router/testing';
import { ActivatedRoute } from '@angular/router';
import {Mode} from '../../util/types';

describe('ChipsComponent', () => {
  @Component({
    selector: 'app-chips-spec',
    template: `
      <app-chips>
        <app-chip
          [note]="'a'"
          [noteExtender]="'natural'"
          [mode]="Mode.ionian"
          [selectedColor]="selectedColor">One
        </app-chip>
        <app-chip
          [note]="'b'"
          [noteExtender]="'flat'"
          [mode]="Mode.mixolydian"
          [rounded]="true">Two
        </app-chip>
        <app-chip
          [note]="'f'"
          [noteExtender]="'sharp'"
          [mode]="Mode.minorPentatonic"
        >Two
        </app-chip>
      </app-chips>
    `
  })
  class ChipsSpecComponent {
    Mode = Mode;
    selectedColor: SelectedColor = SelectedColor.default;
  }

  let component: ChipsSpecComponent;
  let fixture: ComponentFixture<ChipsSpecComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        ChipsSpecComponent
      ],
      imports: [
        ChipsModule,
        RouterTestingModule
      ],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            snapshot: {
              data: {
                selected: { note: 'c', noteExtender: 'natural', mode: 'ionian' }
              }
            }
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ChipsSpecComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
