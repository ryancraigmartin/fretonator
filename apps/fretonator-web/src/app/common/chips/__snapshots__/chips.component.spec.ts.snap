// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChipsComponent should match the snapshot 1`] = `
<app-chips-spec
  Mode={[Function Object]}
  __ngContext__={[Function LRootView]}
  selectedColor={[Function String]}
>
  <app-chips>
    <div
      class="chips"
    >
      <app-chip>
        <a
          aria-label="A Ionian (Major)"
          class="chip"
          href="/a/natural/ionian"
        >
          One 
        </a>
      </app-chip>
      <app-chip>
        <a
          aria-label="B flat Mixolydian"
          class="chip chip--round"
          href="/b/flat/mixolydian"
        >
          Two 
        </a>
      </app-chip>
      <app-chip>
        <a
          aria-label="F sharp Minor Pentatonic"
          class="chip"
          href="/f/sharp/minorPentatonic"
        >
          Two 
        </a>
      </app-chip>
    </div>
  </app-chips>
</app-chips-spec>
`;
